File: sdk/samples/WFSamples/Technologies/CustomActivities/FileWatcher/CS/FileWatcherService/FileWatcherService.cs
//---------------------------------------------------------------------
//  This file is part of the Windows Workflow Foundation SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//---------------------------------------------------------------------

using System;
using System.IO;
using System.Collections.Generic;
using System.Workflow.Runtime;

namespace Microsoft.Samples.Workflow.FileWatcher
{
    public class FileWatcherService
    {
        private Dictionary<string, FileWatcherSubscription> subscriptions;
        private WorkflowRuntime runtime;

        public FileWatcherService(WorkflowRuntime runtime)
        {
            this.runtime = runtime;
            this.subscriptions = new Dictionary<string, FileWatcherSubscription>();
        }

        public Guid RegisterListener(IComparable queueName, string path, string filter, NotifyFilters notifyFilter, bool includeSubdirectories)
        {
            FileSystemWatcher fileSystemWatcher = new FileSystemWatcher();

            try
            {
                fileSystemWatcher.Path = path;
            }
            catch (ArgumentException)
            {
                Console.WriteLine("Path \"{0}\" not found.  Please see documentation for setup steps.", path);
                return Guid.Empty;
            }
            fileSystemWatcher.Filter = filter;
            fileSystemWatcher.NotifyFilter = notifyFilter;
            fileSystemWatcher.IncludeSubdirectories = includeSubdirectories;

            fileSystemWatcher.Changed += new FileSystemEventHandler(FileSystemWatcher_Handler);
            fileSystemWatcher.Created += new FileSystemEventHandler(FileSystemWatcher_Handler);
            fileSystemWatcher.Deleted += new FileSystemEventHandler(FileSystemWatcher_Handler);
            fileSystemWatcher.Error += new ErrorEventHandler(FileSystemWatcher_Error);
            
            FileWatcherSubscription subscription =
                new FileWatcherSubscription(fileSystemWatcher, WorkflowEnvironment.WorkflowInstanceId, queueName);

            Guid subscriptionId = Guid.NewGuid();

            lock (this.subscriptions)
            {
                this.subscriptions.Add(subscriptionId.ToString(), subscription);
            }

            // Turn the file system watcher on
            fileSystemWatcher.EnableRaisingEvents = true;

            Console.WriteLine("FileWatcherService subscription '" + subscriptionId.ToString() + "' created");

            return subscriptionId;
        }

        public void UnregisterListener(Guid subscriptionId)
        {
            string key = subscriptionId.ToString();
            lock (this.subscriptions)
            {
                if (this.subscriptions.ContainsKey(key))
                {
                    FileSystemWatcher watcher = this.subscriptions[key].FileSystemWatcher;

                    // Turn the file system watcher off
                    watcher.EnableRaisingEvents = false;
                    watcher.Dispose();

                    this.subscriptions.Remove(key);
                    Console.WriteLine("FileWatcherService subscription '" + key + "' removed");
                }
                else
                {
                    Console.WriteLine("FileWatcherService subscription '" + key + "' not found");
                }
            }
        }

        void FileSystemWatcher_Error(object sender, ErrorEventArgs e)
        {
            
        }

        void FileSystemWatcher_Handler(object sender, FileSystemEventArgs e)
        {
            // Sender is the FileSystemWatcher that raised this event
            FileSystemWatcher fileSystemWatcher = (FileSystemWatcher) sender;
            lock (this.subscriptions)
            {
                foreach (FileWatcherSubscription subscription in this.subscriptions.Values)
                {
                    if (subscription.FileSystemWatcher.Equals(fileSystemWatcher))
                    {
                        this.DeliverToWorkflow(subscription, e);
                        return;
                    }
                }
            }
        }

        private void DeliverToWorkflow(FileWatcherSubscription subscription, FileSystemEventArgs fileSystemEventArgs)
        {
            try
            {
                // We can't just use the FileSystemEventArgs because it's not serializable
                FileWatcherEventArgs eventArgs = new FileWatcherEventArgs(fileSystemEventArgs);

                WorkflowInstance workflowInstance = this.runtime.GetWorkflow(subscription.WorkflowInstanceId);
                workflowInstance.EnqueueItem(subscription.QueueName, eventArgs, null, null);
            }
            catch (Exception e)
            {
                // Write the exception out to the Debug console and throw the exception
                System.Diagnostics.Debug.WriteLine(e);
                throw e;
            }
        }
    }
}

File: sdk/samples/WFSamples/Technologies/CustomActivities/FileWatcher/VB/FileWatcherService/FileWatcherService.vb
'---------------------------------------------------------------------
'  This file is part of the Windows Workflow Foundation SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'---------------------------------------------------------------------

Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Workflow.Runtime

Public Class FileWatcherService
    Private subscriptions As Dictionary(Of String, FileWatcherSubscription)
    Private runtime As WorkflowRuntime


    Public Sub New(ByVal runtime As WorkflowRuntime)
        Me.runtime = runtime
        Me.subscriptions = New Dictionary(Of String, FileWatcherSubscription)()
    End Sub

    Public Function RegisterListener(ByVal queueName As IComparable, ByVal path As String, ByVal filter As String, ByVal notifyFilter As NotifyFilters, ByVal includeSubdirectories As Boolean) As Guid
        Dim fileSystemWatcher As New FileSystemWatcher()

        Try
            fileSystemWatcher.Path = path
        Catch ex As ArgumentException
            Console.WriteLine("Path '{0}' not found.  Please see documentation for setup steps.", path)
            Return Guid.Empty
        End Try

        fileSystemWatcher.Filter = filter
        fileSystemWatcher.NotifyFilter = notifyFilter
        fileSystemWatcher.IncludeSubdirectories = includeSubdirectories

        AddHandler fileSystemWatcher.Changed, AddressOf FileSystemWatcherHandler
        AddHandler fileSystemWatcher.Created, AddressOf FileSystemWatcherHandler
        AddHandler fileSystemWatcher.Deleted, AddressOf FileSystemWatcherHandler
        AddHandler fileSystemWatcher.Error, AddressOf FileSystemWatcherError

        Dim subscription As New FileWatcherSubscription(fileSystemWatcher, WorkflowEnvironment.WorkflowInstanceId, queueName)
        Dim subscriptionId As Guid = Guid.NewGuid()

        SyncLock Me.subscriptions
            Me.subscriptions.Add(subscriptionId.ToString(), subscription)
        End SyncLock

        ' Turn the file system watcher on
        fileSystemWatcher.EnableRaisingEvents = True

        Console.WriteLine("FileWatcherService subscription '" + subscriptionId.ToString() + "' created")

        Return subscriptionId
    End Function

    Public Sub UnregisterListener(ByVal subscriptionId As Guid)
        Dim key As String = subscriptionId.ToString()

        SyncLock Me.subscriptions
            If Me.subscriptions.ContainsKey(key) Then
                Dim watcher As FileSystemWatcher = Me.subscriptions(key).FileSystemWatcher
                ' Turn the file system watcher off
                watcher.EnableRaisingEvents = False
                watcher.Dispose()

                Me.subscriptions.Remove(key)
                Console.WriteLine("FileWatcherService subscription '" + key + "' removed")
            Else
                Console.WriteLine("FileWatcherService subscription '" + key + "' not found")
            End If
        End SyncLock

    End Sub

    Private Sub FileSystemWatcherError(ByVal sender As Object, ByVal e As ErrorEventArgs)

    End Sub

    Private Sub FileSystemWatcherHandler(ByVal sender As Object, ByVal e As FileSystemEventArgs)
        ' Sender is the FileSystemWatcher that raised this event
        Dim fileSystemWatcher As FileSystemWatcher = CType(sender, FileSystemWatcher)
        SyncLock Me.subscriptions
            For Each subscription As FileWatcherSubscription In Me.subscriptions.Values
                If subscription.FileSystemWatcher.Equals(fileSystemWatcher) Then
                    Me.DeliverToWorkflow(subscription, e)
                    Return

                End If
            Next
        End SyncLock
    End Sub

    Private Sub DeliverToWorkflow(ByVal subscription As FileWatcherSubscription, ByVal fileSystemEventArgs As FileSystemEventArgs)
        Try
            ' We can't just use the FileSystemEventArgs because it's not serializable
            Dim eventArgs As New FileWatcherEventArgs(fileSystemEventArgs)

            Dim workflowInstance As WorkflowInstance = Me.runtime.GetWorkflow(subscription.WorkflowInstanceId)
            workflowInstance.EnqueueItem(subscription.QueueName, eventArgs, Nothing, Nothing)
        Catch e As Exception
            ' Write the exception out to the Debug console and throw the exception
            System.Diagnostics.Debug.WriteLine(e)
            Throw e
        End Try
    End Sub

End Class
